hera_cal/abscal.py

```diff
@@ -3693,6 +3694,158 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        Default is None -> automatically select a refant.
+    ant_threshold: float, optional
+        threshold of flags in frequency and time for a given antenna to completely
+        flag this antenna rom calibration.
+    verbose: bool, optional
+        lots of outputs.
+    """
+    hdd = io.HERAData(data_file)
+    hdm = io.HERAData(model_file)
+
+    # expand hdm by redundancy
+    _, data_flags, _ = hdd.read()
+    hdm.read()
+    if inflate_model_by_redundancy:
+        if constrain_model_to_data_ants:
+            # In order to avoid inflating to full interferometer dataset
+            # which will be unecessary for many earlier analyses
+            # prune model antennas to only include data antennas in hdd
+            # this will only work if theantennas in the baseline keys for each redundant group
+            # in the model are also present in the dataset so only use this if you are confident
+            # that this is the case.
+            all_data_ants = np.unique(np.hstack([hdd.ant_1_array, hdd.ant_2_array]))
+            all_model_ants = np.unique(np.hstack([hdm.ant_1_array, hdm.ant_2_array]))
+            assert np.all([ant in all_data_ants for ant in all_model_ants]), "All model antennas must be present in data if constrain_model_to_data_ants is True!"
+            hdm.select(antenna_nums=all_data_ants,
+                       keep_all_metadata=False)
+        hdm.inflate_by_redundancy()
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
+
+    _, model_flags, _ = hdm.build_datacontainers()
+
+    data_flags = synthesize_ant_flags(data_flags, threshold=ant_threshold)
+    model_flags = synthesize_ant_flags(model_flags, threshold=ant_threshold)
+
+    lst_center = hdd.lsts[hdd.Ntimes // 2] * 12 / np.pi
+    field_str = 'LST={lst_center:%.2f} hrs'
+
+    if refant is None:
+        # dummy refant. Could be completely flagged for all we know.
+        refant_init = str(hdd.ant_1_array[0])
+    else:
+        refant_init = str(refant)
+    # initialize HERACal to store new cal solutions
+    hc = UVCal()
+    hc = hc.initialize_from_uvdata(uvdata=hdd, gain_convention='divide', cal_style='sky',
+                                   ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                   metadata_only=False, sky_field=field_str, cal_type='gain')
+    hc.write_calfits(output_filename, clobber=clobber)
+    hc = io.HERACal(output_filename)
+    hc.read()
+    hc.update(flags=data_flags)
+    # generate cal object from model to hold model flags.
+    hcm = UVCal()
+    hcm = hcm.initialize_from_uvdata(uvdata=hdm, gain_convention='divide', cal_style='sky',
+                                     ref_antenna_name=refant_init, sky_catalog=f'{model_file}',
+                                     metadata_only=False, sky_field=field_str, cal_type='gain')
+    hcm.write_calfits(output_filename + '.tmp')
```

Comment:
why do we have a temporary file being written and then removed?

Reply:
The only way I know to initialize a HERACal object is through the filepath to a calfits file -- hence writing the temporary file. If there is a way to initialize HERACal from a uvcal or uvdata/HERAData object in memory this would definitely be the preferable way to go.

Reply:
I think perhaps the function you are looking for is `io.to_HERACal` where you can feed a UVCal object rather than a UVCal filename

Reply:
Yep. Thanks for pointing this out!